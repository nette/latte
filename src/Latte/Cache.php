<?php

/**
 * This file is part of the Latte (https://latte.nette.org)
 * Copyright (c) 2008 David Grudl (https://davidgrudl.com)
 */

declare(strict_types=1);

namespace Latte;


/**
 * Caching of compiled templates.
 * @internal
 */
final class Cache
{
	public ?string $directory = null;
	public bool $autoRefresh = true;


	public function loadOrCreate(Engine $engine, string $name): void
	{
		// Solving atomicity to work everywhere is really pain in the ass.
		// 1) We want to do as little as possible IO calls on production and also directory and file can be not writable
		// so on Linux we include the file directly without shared lock, therefore, the file must be created atomically by renaming.
		// 2) On Windows file cannot be renamed-to while is open (ie by include), so we have to acquire a lock.
		$file = $engine->getCacheFile($name);
		$signature = $this->autoRefresh
			? md5(serialize($this->generateSignature($engine, $name)))
			: null;
		$lock = defined('PHP_WINDOWS_VERSION_BUILD') || $signature
			? $this->acquireLock("$file.lock", LOCK_SH)
			: null;

		if (
			!($signature && $signature !== stream_get_contents($lock))
			&& (@include $file) !== false // @ - file may not exist
		) {
			return;
		}

		if ($lock) {
			flock($lock, LOCK_UN); // release shared lock so we can get exclusive
			fseek($lock, 0);
		}

		$lock = $this->acquireLock("$file.lock", LOCK_EX);

		// while waiting for exclusive lock, someone might have already created the cache
		if (!is_file($file) || ($signature && $signature !== stream_get_contents($lock))) {
			$compiled = $engine->compile($name);
			if (
				file_put_contents("$file.tmp", $compiled) !== strlen($compiled)
				|| !rename("$file.tmp", $file)
			) {
				@unlink("$file.tmp"); // @ - file may not exist
				throw new RuntimeException("Unable to create '$file'.");
			}

			fseek($lock, 0);
			fwrite($lock, $signature ?? md5(serialize($this->generateSignature($engine, $name))));
			ftruncate($lock, ftell($lock));

			if (function_exists('opcache_invalidate')) {
				@opcache_invalidate($file, true); // @ can be restricted
			}
		}

		if ((include $file) === false) {
			throw new RuntimeException("Unable to load '$file'.");
		}

		flock($lock, LOCK_UN);
	}


	/** @return resource */
	private function acquireLock(string $file, int $mode)
	{
		$dir = dirname($file);
		if (!is_dir($dir) && !@mkdir($dir) && !is_dir($dir)) { // @ - dir may already exist
			throw new RuntimeException("Unable to create directory '$dir'. " . error_get_last()['message']);
		}

		$handle = @fopen($file, 'c+'); // @ is escalated to exception
		if (!$handle) {
			throw new RuntimeException("Unable to create file '$file'. " . error_get_last()['message']);
		} elseif (!@flock($handle, $mode)) { // @ is escalated to exception
			throw new RuntimeException('Unable to acquire ' . ($mode & LOCK_EX ? 'exclusive' : 'shared') . " lock on file '$file'. " . error_get_last()['message']);
		}

		return $handle;
	}


	public function generateFileName(string $name, string $hash): string
	{
		$base = preg_match('#([/\\\][\w@.-]{3,35}){1,3}$#D', '/' . $name, $m)
			? preg_replace('#[^\w@.-]+#', '-', substr($m[0], 1)) . '--'
			: '';
		return $this->directory . '/' . $base . $hash . '.php';
	}


	/**
	 * Values that check the expiration of the compiled template.
	 */
	protected function generateSignature(Engine $engine, string $name): array
	{
		return [
			Engine::Version,
			$engine->getLoader()->getContent($name),
			array_map(
				fn($extension) => filemtime((new \ReflectionObject($extension))->getFileName()),
				$engine->getExtensions(),
			),
		];
	}
}
