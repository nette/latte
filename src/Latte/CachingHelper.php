<?php

/**
 * This file is part of the Nette Framework.
 *
 * Copyright (c) 2004, 2011 David Grudl (http://davidgrudl.com)
 *
 * This source file is subject to the "Nette license", and/or
 * GPL license. For more information please see http://nette.org
 */

namespace Nette\Templates;

use Nette,
	Nette\Environment,
	Nette\Caching\Cache;



/**
 * Caching template helper.
 *
 * @author     David Grudl
 */
class CachingHelper extends Nette\Object
{
	/** @var array */
	private $frame;

	/** @var string */
	private $key;



	/**
	 * Starts the output cache. Returns CachingHelper object if buffering was started.
	 * @param  string
	 * @param  array of CachingHelper
	 * @param  array
	 * @return CachingHelper
	 */
	public static function create($key, & $parents, $args = NULL)
	{
		if ($args) {
			if (array_key_exists('if', $args) && !$args['if']) {
				return $parents[] = new self;
			}
			$key = array_merge(array($key), array_intersect_key($args, range(0, count($args))));
		}
		if ($parents) {
			end($parents)->frame[Cache::ITEMS][] = $key;
		}

		$cache = self::getCache();
		if (isset($cache[$key])) {
			echo $cache[$key];
			return FALSE;

		} else {
			$obj = new self;
			$obj->key = $key;
			$obj->frame = array(
				Cache::TAGS => isset($args['tags']) ? $args['tags'] : NULL,
				Cache::EXPIRATION => isset($args['expire']) ? $args['expire'] : '+ 7 days',
			);
			ob_start();
			return $parents[] = $obj;
		}
	}



	/**
	 * Stops and saves the cache.
	 * @return void
	 */
	public function save()
	{
		if ($this->key !== NULL) {
			$this->getCache()->save($this->key, ob_get_flush(), $this->frame);
		}
		$this->key = $this->frame = NULL;
	}



	/**
	 * Adds the file dependency.
	 * @param  string
	 * @return void
	 */
	public function addFile($file)
	{
		$this->frame[Cache::FILES][] = $file;
	}



	/********************* backend ****************d*g**/



	/**
	 * @return Nette\Caching\Cache
	 */
	protected static function getCache()
	{
		return Environment::getCache('Nette.Template.Cache');
	}

}