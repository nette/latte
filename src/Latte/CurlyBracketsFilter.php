<?php

/**
 * Nette Framework
 *
 * Copyright (c) 2004, 2009 David Grudl (http://davidgrudl.com)
 *
 * This source file is subject to the "Nette license" that is bundled
 * with this package in the file license.txt.
 *
 * For more information please see http://nettephp.com
 *
 * @copyright  Copyright (c) 2004, 2009 David Grudl
 * @license    http://nettephp.com/license  Nette license
 * @link       http://nettephp.com
 * @category   Nette
 * @package    Nette\Templates
 * @version    $Id$
 */

/*namespace Nette\Templates;*/



require_once dirname(__FILE__) . '/../../Object.php';



/**
 * Compile-time filter CurlyBrackets.
 *
 * @author     David Grudl
 * @copyright  Copyright (c) 2004, 2009 David Grudl
 * @package    Nette\Templates
 */
class CurlyBracketsFilter extends /*Nette\*/Object
{
	/** single & double quoted PHP string */
	const RE_STRING = '\'(?:\\\\.|[^\'\\\\])*\'|"(?:\\\\.|[^"\\\\])*"';

	/** PHP identifier */
	const RE_IDENTIFIER = '[_a-zA-Z\x7F-\xFF][_a-zA-Z0-9\x7F-\xFF]*';

	/** PHP identifier */
	const RE_CURLY = '
		(?P<indent>\n[ \t]*)?
		\\{(?P<macro>[^\\s\'"{}](?>
			\'(?:\\\\.|[^\'\\\\])*\'|"(?:\\\\.|[^"\\\\])*"|  # RE_STRING
			[^\'"}]+)*)\\}
		(?P<newline>[\ \t]*(?=\r|\n))?
	';

	/** @var array */
	private $handlers;

	/** @var string */
	private $input, $output;

	/** @var int */
	private $offset;

	/** @var strng (for CONTEXT_ATTRIBUTE) */
	private $quote;

	/** @var array */
	private $tags;

	/** @var string */
	public $context, $escape;

	/**#@+ Context-aware escaping states */
	const CONTEXT_TEXT = 'text';
	const CONTEXT_CDATA = 'cdata';
	const CONTEXT_TAG = 'tag';
	const CONTEXT_ATTRIBUTE = 'attribute';
	const CONTEXT_NONE = 'none';
	const CONTEXT_COMMENT = 'comment';
	/**#@-*/



	/**
	 * Adds a new macro handler.
	 * @param  ICurlyBracketsHandler
	 * @return void
	 */
	public function addHandler($handler)
	{
		$this->handlers[] = $handler;
	}



	/**
	 * Invokes filter.
	 * @param  string
	 * @return string
	 */
	public function __invoke($s)
	{
		// context-aware escaping
		$this->context = CurlyBracketsFilter::CONTEXT_NONE;
		$this->escape = '$template->escape';

		// initialize handlers
		if (empty($this->handlers)) {
			$this->handlers[] = new CurlyBracketsMacros;
		}
		foreach ($this->handlers as $handler) {
			$handler->initialize($this, $s);
		}

		// process all {tags} and <tags/>
		$s = $this->parse("\n" . $s);

		foreach ($this->handlers as $handler) {
			$handler->finalize($s);
		}

		return $s;
	}



	/**
	 * Searches for curly brackets, HTML tags and attributes.
	 * @param  string
	 * @return string
	 */
	private function parse($s)
	{
		$this->input = & $s;
		$this->offset = 0;
		$this->output = '';
		$this->tags = array();
		$len = strlen($s);

		while ($this->offset < $len) {
			$matches = $this->{"context$this->context"}();

			if (!$matches) { // EOF
				break;

			} elseif (!empty($matches['macro'])) { // {macro|modifiers}
				preg_match('#^(/?[a-z]+)?(.*?)(\\|[a-z](?:'.self::RE_STRING.'|[^\'"\s]+)*)?$()#i', $matches['macro'], $m2);
				list(, $macro, $value, $modifiers) = $m2;
				$code = $this->processMacro($macro, trim($value), isset($modifiers) ? $modifiers : '');
				if ($code === NULL) {
					throw new /*\*/InvalidStateException("Unknown macro '{$matches['macro']}'.");
				}
				$nl = isset($matches['newline']) ? "\n" : ''; // double newline
				if ($nl && $matches['indent'] && strncmp($code, '<?php echo ', 11)) {
					$this->output .= "\n" . $code; // remove indent, single newline
				} else {
					$this->output .= $matches['indent'] . $code . $nl;
				}

			} else { // common behaviour
				$this->output .= $matches[0];
			}
		}

		return $this->output . substr($this->input, $this->offset);
	}



	/**
	 * Handles CONTEXT_TEXT.
	 */
	private function contextText()
	{
		$matches = $this->match('~
			<(?P<closing>/?)(?P<tag>[a-z0-9:]+)|  ##  begin of HTML tag <tag </tag - ignores <!DOCTYPE
			<(?P<comment>!--)|           ##  begin of HTML comment <!--
			'.self::RE_CURLY.'           ##  curly tag
		~xsi');

		if (!$matches || !empty($matches['macro'])) { // EOF or {macro}

		} elseif (!empty($matches['comment'])) { // <!--
			$this->context = self::CONTEXT_COMMENT;
			$this->escape = 'TemplateHelpers::escapeHtmlComment';

		} elseif (empty($matches['closing'])) { // <tag
			$tag = $this->tags[] = (object) NULL;
			$tag->name = $matches['tag'];
			$tag->closing = FALSE;
			$this->context = self::CONTEXT_TAG;
			$this->escape = 'TemplateHelpers::escapeHtml';

		} else { // </tag
			$tag = end($this->tags);
			$tag->name = $matches['tag'];
			$tag->closing = TRUE;
			$this->context = self::CONTEXT_TAG;
			$this->escape = 'TemplateHelpers::escapeHtml';
		}
		return $matches;
	}



	/**
	 * Handles CONTEXT_CDATA.
	 */
	private function contextCData()
	{
		$tag = end($this->tags);
		$matches = $this->match('~
			</'.$tag->name.'(?![a-z0-9:])| ##  end HTML tag </tag
			'.self::RE_CURLY.'           ##  curly tag
		~xsi');

		if ($matches && empty($matches['macro'])) { // </tag
			$tag->closing = TRUE;
			$this->context = self::CONTEXT_TAG;
			$this->escape = 'TemplateHelpers::escapeHtml';
		}
		return $matches;
	}



	/**
	 * Handles CONTEXT_TAG.
	 */
	private function contextTag()
	{
		$matches = $this->match('~
			(?P<end>>)|                  ##  end of HTML tag
			(?<=\\s)(?P<attr>[a-z0-9:-]+)\s*=\s*(?P<quote>["\'])| ## begin of HTML attribute
			'.self::RE_CURLY.'           ##  curly tag
		~xsi');

		if (!$matches || !empty($matches['macro'])) { // EOF or {macro}

		} elseif (!empty($matches['end'])) { // >
			$tag = end($this->tags);

			if (!$tag->closing && (strcasecmp($tag->name, 'script') === 0 || strcasecmp($tag->name, 'style') === 0)) {
				$this->context = self::CONTEXT_CDATA;
				$this->escape = strcasecmp($tag->name, 'style') ? 'TemplateHelpers::escapeJs' : 'TemplateHelpers::escapeCss';
			} else {
				$this->context = self::CONTEXT_TEXT;
				$this->escape = 'TemplateHelpers::escapeHtml';
				if ($tag->closing) array_pop($this->tags);
			}

		} else { // attribute = '"
				$this->context = self::CONTEXT_ATTRIBUTE;
			$this->quote = $matches['quote'];
			$this->escape = strncasecmp($matches['attr'], 'on', 2)
				? (strcasecmp($matches['attr'], 'style') ? 'TemplateHelpers::escapeHtml' : 'TemplateHelpers::escapeHtmlCss')
					: 'TemplateHelpers::escapeHtmlJs';
		}
		return $matches;
	}



	/**
	 * Handles CONTEXT_ATTRIBUTE.
	 */
	private function contextAttribute()
	{
		$matches = $this->match('~
			(' . $this->quote . ')|      ##  1) end of HTML attribute
			'.self::RE_CURLY.'           ##  curly tag
		~xsi');

		if ($matches && empty($matches['macro'])) { // (attribute end) '"
			$this->context = self::CONTEXT_TAG;
			$this->escape = 'TemplateHelpers::escapeHtml';
		}
		return $matches;
	}



	/**
	 * Handles CONTEXT_COMMENT.
	 */
	private function contextComment()
	{
		$matches = $this->match('~
			(--\s*>)|                    ##  1) end of HTML comment
			'.self::RE_CURLY.'           ##  curly tag
		~xsi');

		if ($matches && empty($matches['macro'])) { // --\s*>
			$this->context = self::CONTEXT_TEXT;
			$this->escape = 'TemplateHelpers::escapeHtml';
		}
		return $matches;
	}



	/**
	 * Handles CONTEXT_NONE.
	 */
	private function contextNone()
	{
		$matches = $this->match('~
			.*                           ##  all
		~xsi');
		return $matches;
	}



	/**
	 * Macro processing.
	 * @param  string
	 * @param  string
	 * @param  string
	 * @return string
	 */
	protected function processMacro($macro, $value, $modifiers)
	{
		foreach ($this->handlers as $handler) {
			$code = $handler->macro($macro, $value, $modifiers);
			if ($code !== NULL) return $code;
		}
	}



	/**
	 * Matches next token.
	 * @param  string
	 * @return array
	 */
	private function match($re)
	{
		if (preg_match($re, $this->input, $matches, PREG_OFFSET_CAPTURE, $this->offset)) {
			$this->output .= substr($this->input, $this->offset, $matches[0][1] - $this->offset);
			$this->offset = $matches[0][1] + strlen($matches[0][0]);
			foreach ($matches as $k => $v) $matches[$k] = $v[0];
		}
		return $matches;
	}



	/********************* compile-time helpers ****************d*g**/



	/**
	 * Applies modifiers.
	 * @param  string
	 * @param  string
	 * @return string
	 */
	public static function formatModifiers($var, $modifiers)
	{
		if (!$modifiers) return $var;
		preg_match_all(
			'~
				'.self::RE_STRING.'|  ## single or double quoted string
				[^\'"|:,]+|           ## symbol
				[|:,]                 ## separator
			~xs',
			$modifiers . '|',
			$tokens
		);
		$inside = FALSE;
		$prev = '';
		foreach ($tokens[0] as $token) {
			if ($token === '|' || $token === ':' || $token === ',') {
				if ($prev === '') {

				} elseif (!$inside) {
					if (!preg_match('#^'.self::RE_IDENTIFIER.'$#', $prev)) {
						throw new /*\*/InvalidStateException("Modifier name must be alphanumeric string, '$prev' given.");
					}
					$var = "\$template->$prev($var";
					$prev = '';
					$inside = TRUE;

				} else {
					$var .= ', ' . self::formatString($prev);
					$prev = '';
				}

				if ($token === '|' && $inside) {
					$var .= ')';
					$inside = FALSE;
				}
			} else {
				$prev .= $token;
			}
		}
		return $var;
	}



	/**
	 * Reads single token (optionally delimited by comma) from string.
	 * @param  string
	 * @return string
	 */
	public static function fetchToken(& $s)
	{
		if (preg_match('#^((?>'.self::RE_STRING.'|[^\'"\s,]+)+)\s*,?\s*(.*)$#', $s, $matches)) { // token [,] tail
			$s = $matches[2];
			return $matches[1];
		}
		return NULL;
	}



	/**
	 * Formats parameters to PHP array.
	 * @param  string
	 * @param  string
	 * @return string
	 */
	public static function formatArray($s, $prefix = '')
	{
		$s = preg_replace_callback(
			'~
				'.self::RE_STRING.'|                          ## single or double quoted string
				(?<=[,=(]|=>|^)\s*([a-z\d_]+)(?=\s*[,=)]|$)   ## 1) symbol
			~xi',
			array(__CLASS__, 'cbArgs'),
			trim($s)
		);
		return $s === '' ? '' : $prefix . "array($s)";
	}



	/**
	 * Callback for formatArgs().
	 */
	private static function cbArgs($matches)
	{
		//    [1] => symbol

		if (!empty($matches[1])) { // symbol
			list(, $symbol) = $matches;
			static $keywords = array('true'=>1, 'false'=>1, 'null'=>1, 'and'=>1, 'or'=>1, 'xor'=>1, 'clone'=>1, 'new'=>1);
			return is_numeric($symbol) || isset($keywords[strtolower($symbol)]) ? $matches[0] : "'$symbol'";

		} else {
			return $matches[0];
		}
	}



	/**
	 * Formats parameter to PHP string.
	 * @param  string
	 * @return string
	 */
	public static function formatString($s)
	{
		return (is_numeric($s) || strspn($s, '\'"$')) ? $s : '"' . $s . '"';
	}



	/**
	 * Invokes filter.
	 * @deprecated
	 */
	public static function invoke($s)
	{
		$filter = new self;
		return $filter->__invoke($s);
	}

}
