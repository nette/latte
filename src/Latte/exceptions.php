<?php

/**
 * This file is part of the Latte (https://latte.nette.org)
 * Copyright (c) 2008 David Grudl (https://davidgrudl.com)
 */

declare(strict_types=1);

namespace Latte;


/**
 * The exception occured during Latte compilation.
 */
class CompileException extends \Exception
{
	/** @var string */
	public $sourceCode;

	/** @var string */
	public $sourceName;

	/** @var int */
	public $sourceLine;


	public function setSource(string $code, int $line, string $name = null)
	{
		$this->sourceCode = $code;
		$this->sourceLine = $line;
		$this->sourceName = $name;
		if (@is_file($name)) { // @ - may trigger error
			$this->message = rtrim($this->message, '.')
				. ' in ' . str_replace(dirname(dirname($name)), '...', $name) . ($line ? ":$line" : '');
		}
		return $this;
	}
}


/**
 * The exception that indicates error of the last Regexp execution.
 */
class RegexpException extends \Exception
{
	public const MESSAGES = [
		PREG_INTERNAL_ERROR => 'Internal error',
		PREG_BACKTRACK_LIMIT_ERROR => 'Backtrack limit was exhausted',
		PREG_RECURSION_LIMIT_ERROR => 'Recursion limit was exhausted',
		PREG_BAD_UTF8_ERROR => 'Malformed UTF-8 data',
		PREG_BAD_UTF8_OFFSET_ERROR => 'Offset didn\'t correspond to the begin of a valid UTF-8 code point',
		6 => 'Failed due to limited JIT stack space', // PREG_JIT_STACKLIMIT_ERROR
	];


	public function __construct($message, $code = null)
	{
		parent::__construct($message ?: (self::MESSAGES[$code] ?? 'Unknown error'), $code);
	}
}


/**
 * Exception thrown when a not allowed construction is used in a template.
 */
class SecurityViolation extends \Exception
{
}
