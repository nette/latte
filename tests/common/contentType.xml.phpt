<?php

/**
 * Test: {contentType application/xml}
 */

declare(strict_types=1);

use Latte\ContentType;
use Latte\Runtime\Html;
use Tester\Assert;

require __DIR__ . '/../bootstrap.php';


$latte = new Latte\Engine;
$latte->setLoader(new Latte\Loaders\StringLoader);

Assert::same(
	' &lt;&gt;&quot;&apos;',
	$latte->renderToString('{contentType xml} {$foo}', ['foo' => '<>"\'']),
);

Assert::same(
	' &lt;&gt;&quot;&apos;',
	$latte->renderToString('{contentType application/xml} {$foo}', ['foo' => '<>"\'']),
);


$latte = new Latte\Engine;
$latte->setContentType(ContentType::Xml);

$params['hello'] = '<i>Hello</i>';
$params['id'] = ':/item';
$params['people'] = ['John', 'Mary', 'Paul', ']]> <!--'];
$params['comment'] = 'test -- comment';
$params['el'] = new Html("<div title='1/2\"'></div>");
$params['xss'] = 'some&<>"\'/chars';

Assert::matchFile(
	__DIR__ . '/expected/contentType.xml.php',
	$latte->compile(__DIR__ . '/templates/contentType.xml.latte'),
);
Assert::matchFile(
	__DIR__ . '/expected/contentType.xml.html',
	$latte->renderToString(
		__DIR__ . '/templates/contentType.xml.latte',
		$params,
	),
);
