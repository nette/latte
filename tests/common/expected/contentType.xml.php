<?php
%A%
final class Template%a% extends Latte\Runtime\Template
{
	public const ContentType = 'xml';


	public function main(array $ʟ_args): void
	{
		extract($ʟ_args);
		unset($ʟ_args);

		echo '<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/css" href="';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo '"?>

<script>';
		if (1) /* line %d% */ {
			echo '<meta />';
		}
		echo '</script>


<ul>
	<li>Escaped: ';
		echo LR\Filters::escapeXml($hello) /* line %d% */;
		echo '</li>
	<li>Non-escaped: ';
		echo $hello /* line %d% */;
		echo '</li>
	<li>Escaped expression: ';
		echo LR\Filters::escapeXml('<' . 'b' . '>hello' . '</b>') /* line %d% */;
		echo '</li>
	<li>Non-escaped expression: ';
		echo '<' . 'b' . '>hello' . '</b>' /* line %d% */;
		echo '</li>
	<li>Array access: ';
		echo LR\Filters::escapeXml($people[1]) /* line %d% */;
		echo '</li>
	<li>Html: ';
		echo LR\Filters::escapeXml($el) /* line %d% */;
		echo '</li>
</ul>

<style type="text/css">
<!--
#';
		echo LR\Filters::escapeHtmlComment($id) /* line %d% */;
		echo ' {
	background: blue;
}
-->
</style>


<script>
<!--
var html = ';
		echo LR\Filters::escapeHtmlComment($el) /* line %d% */;
		echo ';
-->
</script>


<p onclick=\'alert(';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo ');alert("hello");\'
 title=\'';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo '"\'
 style="color:';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo ';\'"
 alt=\'';
		echo LR\Filters::escapeXml($el) /* line %d% */;
		echo '\'
 onfocus="alert(';
		echo LR\Filters::escapeXml($el) /* line %d% */;
		echo ')"
>click on me</p>


<!-- ';
		echo LR\Filters::escapeHtmlComment($comment) /* line %d% */;
		echo ' -->


</ul>


<ul>
';
		foreach ($people as $person) /* line %d% */ {
			echo '	<li>';
			echo LR\Filters::escapeXml($person) /* line %d% */;
			echo '</li>
';

		}

		echo '</ul>

';
		if (true) /* line %d% */ {
			echo '<p>
	<div><p>true</div>
</p>
';
		}
		echo '
<input/> <input />

<p val=';
		if (true) /* line %d% */ {
			echo '"a"';
		} else /* line %d% */ {
			echo '"b"';
		}
		echo '> </p>

<p val=';
		echo '"' . LR\Filters::escapeXml($xss) . '"' /* line %d% */;
		echo ' val2=';
		echo '"' . LR\Filters::escapeXml($mxss) . '"' /* line %d% */;
		echo '> </p>

<p onclick=';
		echo '"' . LR\Filters::escapeXml($xss) . '"' /* line %d% */;
		echo '> </p>
';
	}


	public function prepare(): array
	{
		extract($this->params);

		if (!$this->getReferringTemplate() || $this->getReferenceType() === 'extends') {
			foreach (array_intersect_key(['person' => '50'], $this->params) as $ʟ_v => $ʟ_l) {
				trigger_error("Variable \$$ʟ_v overwritten in foreach on line $ʟ_l");
			}
		}
		if (empty($this->global->coreCaptured) && in_array($this->getReferenceType(), ['extends', null], true)) {
			header('Content-Type: application/xml; charset=utf-8') /* line %d% */;
		}
		return get_defined_vars();
	}
}
