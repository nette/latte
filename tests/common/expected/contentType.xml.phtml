<?php
%A%
final class Template%a% extends Latte\Runtime\Template
{
	protected const CONTENT_TYPE = 'xml';


	public function main(): array
	{
		extract($this->params);
		if (empty($this->global->coreCaptured) && in_array($this->getReferenceType(), ['extends', null], true)) {
			header('Content-Type: application/xml; charset=utf-8') /* line %d% */;
		}
		echo '<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/css" href="';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo '"?>

<script>';
		if (1) /* line %d% */ {
			echo '<meta/>';
		}
		echo '</script>


<ul>
	<li>Escaped: ';
		echo LR\Filters::escapeXml($hello) /* line %d% */;
		echo '</li>
	<li>Non-escaped: ';
		echo $hello /* line %d% */;
		echo '</li>
	<li>Escaped expression: ';
		echo LR\Filters::escapeXml('<' . 'b' . '>hello' . '</b>') /* line %d% */;
		echo '</li>
	<li>Non-escaped expression: ';
		echo '<' . 'b' . '>hello' . '</b>' /* line %d% */;
		echo '</li>
	<li>Array access: ';
		echo LR\Filters::escapeXml($people[1]) /* line %d% */;
		echo '</li>
	<li>Html: ';
		echo LR\Filters::escapeXml($el) /* line %d% */;
		echo '</li>
</ul>

<style type="text/css">
<!--
#';
		echo LR\Filters::escapeHtmlComment($id) /* line %d% */;
		echo ' {
	background: blue;
}
-->
</style>


<script>
<!--
var html = ';
		echo LR\Filters::escapeHtmlComment($el) /* line %d% */;
		echo ';
-->
</script>


<p onclick =
\'alert(';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo ');alert("hello");\'
 title=\'';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo '"\'
 style =
 "color:';
		echo LR\Filters::escapeXml($id) /* line %d% */;
		echo ';\'"
 alt=\'';
		echo LR\Filters::escapeXml($el) /* line %d% */;
		echo '\'
 onfocus="alert(';
		echo LR\Filters::escapeXml($el) /* line %d% */;
		echo ')">click on me</p>


<!-- ';
		echo LR\Filters::escapeHtmlComment($comment) /* line %d% */;
		echo ' -->


</ul>


<ul>
';
		$iterations = 0;
		foreach ($people as $person) /* line %d% */ {
			echo '	<li>';
			echo LR\Filters::escapeXml($person) /* line %d% */;
			echo '</li>
';
			$iterations++;
		}

		echo '</ul>

';
		if (true) /* line %d% */ {
			echo '<p>
	<div><p>true</div>
</p>
';
		}
		echo '
<input/> <input/>

<p val = ';
		if (true) /* line %d% */ {
			echo '"a"';
		} else /* line %d% */ {
			echo '"b"';
		}
		echo '> </p>

<p val = ';
		echo LR\Filters::escapeXmlAttrUnquoted($xss) /* line %d% */;
		echo ' val2=';
		echo LR\Filters::escapeXmlAttrUnquoted($mxss) /* line %d% */;
		echo '> </p>

<p onclick = ';
		echo LR\Filters::escapeXmlAttrUnquoted($xss) /* line %d% */;
		echo '> </p>

<p val = />';
		echo LR\Filters::escapeXml($xss) /* line %d% */;
		echo '</p>
';
		return get_defined_vars();
	}


	public function prepare(): void
	{
		extract($this->params);
		if (!$this->getReferringTemplate() || $this->getReferenceType() === "extends") {
			foreach (array_intersect_key(['person' => '50'], $this->params) as $ʟ_v => $ʟ_l) {
				trigger_error("Variable \$$ʟ_v overwritten in foreach on line $ʟ_l");
			}
		}
	}

}
