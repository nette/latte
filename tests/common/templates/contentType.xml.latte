{contentType application/xml; charset=utf-8}
<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/css" href="{$id}"?>

<script><meta n:if=1 /></script>


<ul>
	<li>Escaped: {$hello}</li>
	<li>Non-escaped: {$hello|noescape}</li>
	<li>Escaped expression: {='<' . 'b' . '>hello' . '</b>'}</li>
	<li>Non-escaped expression: {='<' . 'b' . '>hello' . '</b>'|noescape}</li>
	<li>Array access: {$people[1]}</li>
	<li>Html: {$el}</li>
</ul>

<style type="text/css">
<!--
#{$id} {
	background: blue;
}
-->
</style>


<script>
<!--
var html = {$el};
-->
</script>


<p onclick =
'alert({$id});alert("hello");'
 title='{$id}"'
 style =
 "color:{$id};'"
 alt='{$el}'
 onfocus="alert({$el})"
>click on me</p>


<!-- {$comment} -->


</ul>


<ul>
	<li n:foreach="$people as $person">{$person}</li>
</ul>

<p n:if="true">
	<div><p>true</div>
</p>

<input/> <input />

<p val = {if true}"a"{else}"b"{/if}> </p>

<p val = {$xss} val2={$mxss}> </p>

<p onclick = {$xss}> </p>

<p title="{="foo a='a' b=\"b\">"|noescape}"></p>
<p {="foo a='a' b=\"b\">"|noescape}></p>
