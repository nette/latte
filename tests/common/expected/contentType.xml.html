<?xml version="1.0" encoding="utf-8"?>
<?xml-stylesheet type="text/css" href=":/item"?>

<script><meta /></script>


<ul>
	<li>Escaped: &lt;i&gt;Hello&lt;/i&gt;</li>
	<li>Non-escaped: <i>Hello</i></li>
	<li>Escaped expression: &lt;b&gt;hello&lt;/b&gt;</li>
	<li>Non-escaped expression: <b>hello</b></li>
	<li>Array access: Mary</li>
	<li>Html: <div title='1/2"'></div></li>
</ul>

<style type="text/css">
<!--
#:/item {
	background: blue;
}
-->
</style>


<script>
<!--
var html = <div title='1/2"'></div>;
-->
</script>


<p onclick='alert(:/item);alert("hello");'
 title=':/item"'
 style="color::/item;'"
 alt='<div title='1/2"'></div>'
 onfocus="alert(<div title='1/2"'></div>)"
>click on me</p>


<!-- test - -  comment -->


</ul>


<ul>
	<li>John</li>
	<li>Mary</li>
	<li>Paul</li>
	<li>]]&gt; &lt;!--</li>
</ul>

<p>
	<div><p>true</div>
</p>

<input/> <input />

<p val="a"> </p>

<p val="some&amp;&lt;&gt;&quot;&apos;/chars" val2="`mxss"> </p>

<p onclick="some&amp;&lt;&gt;&quot;&apos;/chars"> </p>

<p title="foo a='a' b=&quot;b&quot;>"></p>
<p foo a='a' b="b">></p>